# A16.2

In [12]:
m = 600
n = 4000
C = randn(m,n) 
d = randn(m)

600-element Vector{Float64}:
 -0.8597495872882819
  1.3266638536917639
  0.1828682546069197
  1.2506364109399535
  2.4473948245701873
 -0.36338535264001415
 -0.27504642905435767
  1.1551394123451821
  1.479712863941644
  0.7124867514178441
 -0.9297246978347928
  0.4357191138873176
 -1.5439696691608136
  ⋮
 -0.13103953512066593
  0.3899732424468207
  1.2090744560052151
  1.4207772757146833
  1.496219916360325
  0.944331523451049
 -0.6799219198642809
  1.2870732127018607
  0.6523879499493365
 -0.1687569925246509
 -0.4449776085970101
 -0.5519537118812325

In [18]:
@time begin
C\d
end

  0.502442 seconds (3.64 k allocations: 41.110 MiB)


4000-element Vector{Float64}:
 -0.0019985706245029934
  0.00693204222180518
  0.0024025505661740883
  0.013160437033790228
  0.0009819741214333977
 -0.0033632268008903024
  0.01533612164737734
  0.0075666387670403415
  0.008671654593281845
  0.00033599862434139944
  0.005390644806690147
  0.0026660520845686593
 -0.0017290434783342281
  ⋮
 -0.0010685522738063773
 -0.0018411258845518995
  0.004930152281968376
  0.01200422672403683
  0.012523705934609712
 -0.0061094315937332425
  0.014272337459122526
 -0.0066142775412051085
 -0.0044998376140026365
 -0.010778202581032608
 -0.003496978352312654
 -0.007800324325818337

In [19]:
@time begin
C\d
end

  0.515546 seconds (3.64 k allocations: 41.110 MiB, 0.21% gc time)


4000-element Vector{Float64}:
 -0.0019985706245029934
  0.00693204222180518
  0.0024025505661740883
  0.013160437033790228
  0.0009819741214333977
 -0.0033632268008903024
  0.01533612164737734
  0.0075666387670403415
  0.008671654593281845
  0.00033599862434139944
  0.005390644806690147
  0.0026660520845686593
 -0.0017290434783342281
  ⋮
 -0.0010685522738063773
 -0.0018411258845518995
  0.004930152281968376
  0.01200422672403683
  0.012523705934609712
 -0.0061094315937332425
  0.014272337459122526
 -0.0066142775412051085
 -0.0044998376140026365
 -0.010778202581032608
 -0.003496978352312654
 -0.007800324325818337

In [20]:
@time begin
C\d
end

  0.507995 seconds (3.64 k allocations: 41.110 MiB, 0.56% gc time)


4000-element Vector{Float64}:
 -0.0019985706245029934
  0.00693204222180518
  0.0024025505661740883
  0.013160437033790228
  0.0009819741214333977
 -0.0033632268008903024
  0.01533612164737734
  0.0075666387670403415
  0.008671654593281845
  0.00033599862434139944
  0.005390644806690147
  0.0026660520845686593
 -0.0017290434783342281
  ⋮
 -0.0010685522738063773
 -0.0018411258845518995
  0.004930152281968376
  0.01200422672403683
  0.012523705934609712
 -0.0061094315937332425
  0.014272337459122526
 -0.0066142775412051085
 -0.0044998376140026365
 -0.010778202581032608
 -0.003496978352312654
 -0.007800324325818337

In [21]:
@time begin
C\d
end

  0.504992 seconds (3.64 k allocations: 41.110 MiB, 0.56% gc time)


4000-element Vector{Float64}:
 -0.0019985706245029934
  0.00693204222180518
  0.0024025505661740883
  0.013160437033790228
  0.0009819741214333977
 -0.0033632268008903024
  0.01533612164737734
  0.0075666387670403415
  0.008671654593281845
  0.00033599862434139944
  0.005390644806690147
  0.0026660520845686593
 -0.0017290434783342281
  ⋮
 -0.0010685522738063773
 -0.0018411258845518995
  0.004930152281968376
  0.01200422672403683
  0.012523705934609712
 -0.0061094315937332425
  0.014272337459122526
 -0.0066142775412051085
 -0.0044998376140026365
 -0.010778202581032608
 -0.003496978352312654
 -0.007800324325818337

In [22]:
@time begin
C\d
end

  0.509473 seconds (3.64 k allocations: 41.110 MiB, 0.54% gc time)


4000-element Vector{Float64}:
 -0.0019985706245029934
  0.00693204222180518
  0.0024025505661740883
  0.013160437033790228
  0.0009819741214333977
 -0.0033632268008903024
  0.01533612164737734
  0.0075666387670403415
  0.008671654593281845
  0.00033599862434139944
  0.005390644806690147
  0.0026660520845686593
 -0.0017290434783342281
  ⋮
 -0.0010685522738063773
 -0.0018411258845518995
  0.004930152281968376
  0.01200422672403683
  0.012523705934609712
 -0.0061094315937332425
  0.014272337459122526
 -0.0066142775412051085
 -0.0044998376140026365
 -0.010778202581032608
 -0.003496978352312654
 -0.007800324325818337

In [24]:
avg_time = (0.502 + 0.515 +0.508 + 0.505 + 0.51)/5

0.508

### Julia took 0.508 seconds on average to solve a least squares problem with  600 equations and n = 4000 variables on my computer.

## Calculation of flop rate:

In [28]:
using GFlops
@count_ops C\d

Flop Counter:
 fma32: 0
 fma64: 0
 add32: 0
 sub32: 0
 mul32: 0
 div32: 0
 add64: 0
 sub64: 1
 mul64: 360001
 div64: 1
 sqrt32: 0
 sqrt64: 1


## For Julia v1.6 the counted number of flops to be 360004 for the backslash least squares operation, so now we can calculate flop rate by taking ratio of number of flops to the average time.

In [35]:
flop_rate = 360004/avg_time

708669.2913385826

## Thus the flop rate is 708669.2913385826 flops/second

# A16.3

## I have considered random values for dimentions of the matrices A,b,C,d since no values have been specified in the problem statement 

In [2]:
m=100
n=250
p=50
A = randn(m,n)
b = randn(m)
C = randn(p,n)
d = randn(p)

50-element Vector{Float64}:
  1.3894335236477646
  1.5191803296492226
  1.127157336391857
 -0.5348184043712568
  2.449829476198982
  0.14156631872600656
 -0.48233947791959914
  2.840652053848935
  1.0686449943076743
 -0.32044369644374765
  0.08907311383310232
  0.31447216847338577
  0.673518659429787
  ⋮
  1.3302202855992526
  1.6102820025494244
  0.22602434506251082
  0.7906818481109996
 -0.9073220530626345
  0.8752681666239364
  2.0774578720248305
 -0.1522112506573425
 -0.4484502648162381
 -0.393291664192649
  1.4128115900619833
 -1.342664700174432

In [3]:
# KKT equation for Constrained Least Squares
xzhat = [2*A'*A C'; C zeros(50,50)] \ [2*A'*b; d] #Since given that associated KKT matrix is invertible
xhat = xzhat[1:250,:]
print(xhat)

[0.18632605256601126; -0.5553945552641091; -1.2519037527232286; 0.44143468222299337; -1.783773830586137; -1.0418810898870206; 0.7558898173167679; -2.3726161634837646; 0.30730975173795944; -2.8959535279826847; 0.38508280241720316; 1.721500220032186; 1.1217421834731176; -1.014101359717818; 0.318589930674954; -0.004197517678421311; 0.7378940563485897; 0.001888390120888961; -0.6939890465848216; 2.2704322396371652; -0.10926503209133856; -0.22144297010425967; 1.8029069408114071; 0.6742598244835829; -2.7462978531145965; 0.8859235996385183; -0.12654188238280403; -0.21233907113508252; 0.9558186183611304; -1.652347246405319; -0.9279474644071924; -0.6995592285138583; 1.729937094177601; 0.5395399219650435; 0.1675476944081075; 0.35259921452192156; -0.8903871928599574; 2.3197407350743884; -1.6779000844171048; 0.6145537982231104; -2.5537768172043096; -2.108481096882038; 1.021097819953129; 2.2312181020174338; -0.4705941622544397; -1.0531614468419104; 1.389595688460653; -0.24131970644072614; 0.47803882